x86: add support for microvm Qemu machine type

francescolavra opened the pull request:
With this change set, the kernel supports booting via the HVM direct boot protocol and running on the microvm Qemu
machine type.
The first 2 commits are fixes for issues in the initialization code run when the kernel is not loaded by the stage2 bootloader.
`make run KERNEL_BOOT=1` runs an image with direct kernel boot (while using the default q35 Qemu machine type); direct kernel boot results in faster boot times compared to bootloader boot.
`make run MICROVM=1` runs an image by using the microvm Qemu machine type; this setting automatically enables direct kernel boot; using microvm results in faster boot times and less memory consumption compared to using the q35 machine type (either with or without direct kernel boot).

sanderssj commented on platform/pc/service.c:
```diff
@@ -501,6 +542,24 @@ void init_service(u64 rdi, u64 rsi)
 }
 
+void pvh_start(hvm_start_info start_info)
+{
```
I tried running this branch on my Mac with KERNEL_BOOT=1, but it crashes here because the function preamble contains the stack smashing guard which references a virtual address that hasn't been mapped yet (e.g. 0xffffffff80131030) and so the unhandled page fault just causes qemu to exit with no messages. I briefly looked for an annotation that can turn it off by function but haven't found one yet. Also not sure if this is just a clang specific problem.

sanderssj commented on platform/pc/service.c:
```diff
@@ -501,6 +542,24 @@ void init_service(u64 rdi, u64 rsi)
 }
 
+void pvh_start(hvm_start_info start_info)
+{
```
Can't reproduce this on any compiler we've tried but the Apple clang 11.0.0. Clang 11.1.0 works fine, and Apple's clang 13.0.0 works fine. This appears to be a compiler bug or quirk that's not applicable to newer versions.

francescolavra commented:
I added support for selecting the microvm machine type in recent Qemu versions (by adding the 'ioapic2=off' flag), and removed the instantiation of the virtio-rng-pci device when microvm is selected (because microvm does not have a PCI bus).

sanderssj reviewed (commented):
LGTM

wjhun reviewed (approved):
LGTM
